subsys/net/lib/lwm2m/lwm2m_engine.c

```diff
@@ -0,0 +1,2473 @@
+			return -EINVAL;
+
+		}
+	}
+
+	if (res->post_write_cb) {
+		/* ignore return value here */
+		res->post_write_cb(obj_inst->obj_inst_id, data_ptr, len,
+				   false, 0);
+	}
+
+	NOTIFY_OBSERVER_PATH(path);
+
+	return 0;
+}
+
+int engine_write_attr_handler(struct lwm2m_engine_obj *obj,
+			      struct lwm2m_engine_context *context)
+{
+	if (!obj || !context) {
+		return -EINVAL;
+	}
+
+	/* TODO: set parameters on resource for notification */
+
+	return 0;
+}
+
+int engine_exec_handler(struct lwm2m_engine_obj *obj,
+			struct lwm2m_engine_context *context)
+{
+	struct lwm2m_obj_path *path = context->path;
+	struct lwm2m_engine_obj_inst *obj_inst;
+	struct lwm2m_engine_res_inst *res = NULL;
+	int ret;
+
+	if (!obj || !context) {
+		return -EINVAL;
+	}
+
+	obj_inst = get_engine_obj_inst(path->obj_id, path->obj_inst_id);
+	if (!obj_inst) {
+		return -ENOENT;
+	}
+
+	ret = engine_get_resource(path, &res);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (res->execute_cb) {
+		res->execute_cb(obj_inst->obj_inst_id);
+		return 0;
+	}
+
+	/* TODO: something else to handle for execute? */
+	return -ENOENT;
+}
+
+int engine_delete_handler(struct lwm2m_engine_obj *obj,
+			  struct lwm2m_engine_context *context)
+{
+	struct lwm2m_obj_path *path = context->path;
+	struct lwm2m_engine_obj_inst *obj_inst;
+
+	SYS_LOG_DBG(">> DELETE [path:%u/%u/%u(%u)]",
+		path->obj_id, path->obj_inst_id, path->res_id, path->level);
+
+	if (!obj || !context) {
+		return -EINVAL;
+	}
+
+	obj_inst = get_engine_obj_inst(path->obj_id, path->obj_inst_id);
+	if (!obj_inst) {
+		return -ENOENT;
+	}
+
+	engine_unregister_obj_inst(obj_inst);
+	obj->instance_count--;
+#ifdef CONFIG_LWM2M_RD_CLIENT_SUPPORT
+	engine_trigger_update();
+#endif
+
+	if (obj->delete_cb) {
+		return obj->delete_cb(path->obj_inst_id);
+	}
+
+	return 0;
+}
+
+#define MATCH_NONE	0
+#define MATCH_ALL	1
+#define MATCH_SINGLE	2
+
+static int do_read_op(struct lwm2m_engine_obj *obj,
+		      struct lwm2m_engine_context *context)
+{
+	struct lwm2m_output_context *out = context->out;
+	struct lwm2m_obj_path *path = context->path;
+	struct lwm2m_engine_obj_inst *obj_inst;
+	int ret = 0, pos = 0, index = 0, match_type;
+	u8_t num_read = 0;
+	u8_t initialized;
+	struct lwm2m_engine_res_inst *res;
+	struct lwm2m_engine_obj_field *obj_field;
+	u16_t temp_res_id;
+
+	obj_inst = get_engine_obj_inst(path->obj_id, path->obj_inst_id);
+	if (!obj_inst) {
+		return -ENOENT;
+	}
+
+	while (obj_inst) {
+		initialized = 0;
+		match_type = MATCH_NONE;
+
+		/* check obj_inst path for at least partial match */
+		if (path->obj_id == obj_inst->obj->obj_id &&
+		    path->obj_inst_id == obj_inst->obj_inst_id) {
+			if (path->level > 2) {
+				match_type = MATCH_SINGLE;
+			} else {
+				match_type = MATCH_ALL;
+			}
+		}
+
+		temp_res_id = path->res_id;
+
+		/* if exact match on obj_inst path, then read all resources */
+		/* if not then read only matching resources */
+
+		if (match_type > MATCH_NONE && obj_inst->resources &&
```

Comment:
ouch, full of if/while/else/... imbrication below, hard to read. (legacy thing again)